### ISSUE TITLE
Rust: Support external crates

### ISSUE THREAD

**Issue by zeenix** [mesonbuild/meson#2173]
I was told that support of Rust was experimental so I can expect hickups and that's understandable but currently it seems external crates are not supported. Since you can't really do much in Rust without external crates, it basically means that Meson doesn't really support Rust. IMO until this issue is addressed in some way, support for Rust shouldn't be advertised at all.

**Comment by TingPing** [comment]
Correct me if I am wrong but Cargo simply just builds all dependencies and there is no concept of external dependencies (ala pkgconfig) right? Does Rust even have any standard directory to install shared libraries? If so we should support `rust.find_library()`.

**Comment by tp-m** [comment]
I think what's needed is more something that does what Cargo does - downloads crates as per `Cargo.toml` and `Cargo.lock` and builds them / links them in as needed. This would be really powerful, esp. for mixed C/Rust projects.

**Comment by TingPing** [comment]
So we either:

- Re-implement Cargo
- Simply call Cargo making the Meson integration worthless

**Comment by zeenix** [comment]
@TingPing typically they just do static linking in Rust currently. `cargo` will simply download, build all deps and then take care of linking your binaries.

**Comment by zeenix** [comment]
@TingPing yeah but please keep in mind that this is useful for projects that will be a mix of Rust and C (or some other language for which meson is a good option).

**Comment by TingPing** [comment]
I realize its valuable but its not a realistic thing we can support. It is like wanting Meson to support building autotools projects. Cargo is its own independent and incompatible build system.

**Comment by zeenix** [comment]
@TingPing I would completely understand if you decide not to support this in meson but then either Rust support shouldn't be advertised at all or this limitation should be explicitly specified. As I mentioned, without this, Rust support is mostly useless.

**Comment by TingPing** [comment]
We have Rust support, we don't have Cargo support. It is sad that the Rust community only cares about Cargo and does not installed shared libraries or have any tooling to collect information about installed libraries.

**Comment by tp-m** [comment]
While it is certainly "independent", It's not clear to me that it really is "incompatible". I see crates more like a special kind of subproject.

**Comment by zeenix** [comment]
@TingPing yeah, i know but fact remains that the support is useless for most uses so really nothing to advertise. I fully agree about Rust folks not doing things the standard way.

**Comment by TingPing** [comment]
I wasn't aware Meson was 'advertising' rust support other than it being documented. It is certainly basic.

**Comment by zeenix** [comment]
IMO basic isn't the correct term. Even the simplest of Rust projects will depend on external crates and meson can't handle those.

**Comment by tp-m** [comment]
I'd say let's focus on what a possible useful solution/implementation would require or could look like, and move the whole what should be advertised or not or whether something useful can be done without this feature or not thing elsewhere :)

**Comment by zeenix** [comment]
@tp-m Sure thing but it seemed like @TingPing is saying "we'll not be doing anything about this" in an earlier comment.

**Comment by nirbheek** [comment]
The main issue seems to be that cargo doesn't just download and manage dependencies, but also builds and links them? If there is no way around that, perhaps we can work with upstream to add a mode that only does the downloading bits.

**Comment by TingPing** [comment]
I would like Meson to be on par with Cargo in features. I don't know how doable it is to make Meson *compatible* with Cargo.

**Comment by TingPing** [comment]
@nirbheek Cargo is effectively a full build system, it downloads dependences, builds them, links them, gets state from the environment, sets variables, etc.

**Comment by mqudsi** [comment]
I think it's hard to appreciate the importance of cargo if you haven't developed in rust.

Here's the problem: unlike almost every other language, the rust "standard library" is purposely as small as possible and covers only what would be necessary to get a basic coding platform that works across multiple operating systems, sort of like a kernel abstraction layer. It doesn't provide much else. Early on in the rust project, code that code be stripped out of rust core and moved to an external package was removed from the standard library. In exchange, certain packages (available via cargo) are 'blessed', developed and maintained by the rust core developers, etc.

Cargo is more than a package manager, it's tightly built into the language itself. Unlike, say GCC or Clang++ with C/C++, rust has evolved with cargo from the start. You write test code inline with the rest of your code, knowing that cargo can pull it out and test it with `cargo test`. You use cargo to manage features in your code and in libraries you depend on (rough equivalent to make defines).

Unlike the situation with C and C++ and their various compilers, rust and cargo can't really be separated. Cargo provides dependency resolution, management, linking, and more, and like meson, has strongly structured contents (in `Cargo.toml`) that guide the build process, and differentiates between libraries and executables, etc.

Cargo is rich enough that it probably has more features than meson currently does. I don't think anyone would view it as a cop out *at all* if meson required cargo and used cargo instead of rustc for all rust interactions. I think it would be insane to try to reproduce the functionality of cargo in meson, and a recipe for disaster.

cargo already has project support, subproject support, source file discovery (provided via the language), and everything else meson would need to (selectively) build a project.

The problem is that simply shelling out to cargo means that dependencies are not defined in cargo.toml but rather in meson.build, but the solution to that is to have meson generate a valid cargo.toml file then run cargo. The structure of cargo.toml is both dumb enough and linear enough (with no dependency resolution *there*) that it would be a joke to write.

So, IMHO, the solution would be to have cargo carry out the work, but meson define the work that needs to be done. I would recommend considering `cargo` one compiler and `rustc` another - if people want a dependency-free, crate-free build they can use the rustc language/compiler and if they want the idiomatic rust experience, the cargo "language" and compiler. `rustc` would take arguments via the command line, while `cargo` would take arguments via a Cargo.toml file that can be generated at build time.

With regards to rebuilding, etc. I think relying on Cargo to do "the right thing" TM would not be unfair. Unlike C/C++, the artifacts of a cargo build are insane.

**Comment by TingPing** [comment]
> Cargo is rich enough that it probably has more features than meson currently does. 

I believe they expose roughly similar features and I think it is also possible to use only `rustc` while retaining those features. The `--cfg` flag seems to be our biggest missing feature currently.

> the solution to that is to have meson generate a valid cargo.toml file then run cargo

Worth noting we would have to implement any support for toml that we need since we can't have dependencies.

I think there is no truely "good" solution as they made their ecosystem self-contained on purpose.

**Comment by mqudsi** [comment]
Thanks for the info. Toml files are just the modern day ini, there's nothing fancy enough to require a dependency (good on you for trying to keep those down to a minimum).

**Comment by TingPing** [comment]
[toml](https://github.com/uiri/toml/blob/master/toml.py) seems to be ~1k LoC, maybe only a subset would be needed but annoying no matter.

**Comment by sdroege** [comment]
Unless you reimplement all of cargo or just call cargo, Rust support in meson is going to be useless. Everything is using lots of other crates and nobody is going to move all of them over to meson.

You *can* call rustc separately but that alone is not going to give you much. Consider cargo the "compiler" and not rustc ;)

You're going to have exactly the same problem with supporting any other language that has a tightly integrated build and dependency tool, and is not a scripting language. Haskell/cabal comes to mind here. Basically any language that did not want to go the C/C++ way of suffering wrt manual building in hundreds of different ways for dependencies.

If you want to become the universal build system, you have to integrate with the build tools of those languages in one way or another.

**Comment by sdroege** [comment]
> toml seems to be ~1k LoC, maybe only a subset would be needed but annoying no matter.

Also toml is just the file format, a more well-defined INI style text format. You'll probably find Python libraries for that, but you'll need multiple times more code to do what cargo does. Also there's "cargo metadata" to parse the toml and do some more things already. But really, don't reimplement all that, it's a waste of time and you'll have to do the same for other languages and then always play catch up

**Comment by jpakkane** [comment]
> Consider cargo the "compiler" and not rustc

But that is not going to work, because Cargo does its own dependency resolution and downloading. Suppose for example that we have a master project that uses a library called `foo`. Assume further that we want to use some Rust code that also uses `foo` via some sort of a Rust wrapper. The important thing to note here is that there must only be _one_ version of `foo` and it _must_ be the same one for both. If you just call into Cargo it will download and build its own crateified version of `foo` (at least this is what I have been told by Rust people that happens) but Meson does not know of this and thus downloads and builds its own version (which it must because things in the build dir are special).

Now we have two versions of `foo` scattered about. In the best possible case the binary is 2x bigger than it should be. In the worst case symbols from these get mixed up and that leads to debugging sprints from hell.

**Comment by sdroege** [comment]
The problem is that you don't even know if too can be reused at all: it might have different, incompatible configurations for example. Cargo knows how to handle that though.

I would consider each Rust crate plus its dependencies one "unit" in meson, and let it output a cdylib (shared lib with only C API exported), staticlib (.a) or rlib (Rust shared lib). And that resulting artefact would then be used further in meson.

Cargo does not have to download things BTW, see "cargo vendor". But it does dependency resolution, build configuration (of dependencies and the top-level crate itself) and the actual build (which might involve not just rustc, but can be arbitrary commands via build.rs, including build of e.g. C code). And docs generation, unit testing, benchmarking, ...

**Comment by mmstick** [comment]
I wanted to use Meson for handling non-Rust aspects of building and installing software, but without Cargo support, Meson is pretty much useless. My project is written 100% in Rust and Cargo works fine to produce a fully static binary with fully reproducible builds (thanks to the Cargo.lock file), but I need a means of checking if the dev box has `mdbook` installed, installing it if it isn't, generating the complete HTML documentation for the project, and offering a means of conveniently installing the project's documentation and other miscellaneous files to the correct locations, and/or automating the process of packaging with certain package managers. Meson doesn't seem up to the task, so I guess I'll just use Ion to script this.

**Comment by sdroege** [comment]
> @sdroege Calling Meson's Rust support useless for not directly calling Cargo is like calling it's C support useless for not calling Autotools directly.

Not really. Autotools is not really a standard for C and as you say instead things are installed as static or shared libraries in a defined way.

With Rust and lots of other languages (this is not really about Rust only!) the whole community decided on a tool for the job and everybody uses it. They're not going to switch to meson. So you would have things to be installed in a defined way, which in the case of Rust (and again, other languages) is also not the case and would have other problems. So as it is right now, the way how meson supports Rust is useless and there is a difference in philosophy that does not make both play well together unless one of them adapts. And again, this is not only Rust that works like that.

The general way of building things there is that you have a leaf-project and that is build as one unit with all its dependencies. For the dependencies the leaf-project could define specific configurations (which make the build results of those dependencies specific to the leaf-project and not reusable) and things like LTO can be performed in the end.


Having a defined way of installing build results would be nice to have, as would be shared libraries with a defined ABI (which would be less about code duplication than in C though, due to generics and monomorphisation, just like in C++). But that's not going to happen anytime soon in either of these languages as the immediate need is not there for anybody using them. Things work fine as-is.

**Comment by sdroege** [comment]
https://github.com/rust-lang/rfcs/pull/2136 This is very relevant here and meson people probably want to get involved into those discussions. Especially the example of bazel given there sounds like more or less how meson wants to work.

**Comment by nirbheek** [comment]
No one linked this here yet, but one approach is: https://github.com/mesonbuild/meson/pull/2617

**Comment by norcalli** [comment]
FYI, as a summary to anyone who doesn't want to read #2617, it is *highly* unlikely that this will be supported. meson won't reimplement cargo and won't integrate with it either. From everything I've read from multiple meson members, they seem to actively dislike Cargo, which is pretty disappointing, but, from their perspective/philosophy, understandable.

**Comment by xclaesse** [comment]
I've been following/reviewing the meson integration of gst-plugins-rs (GStreamer Rust plugins), and I think it has a design that can work. https://gitlab.freedesktop.org/gstreamer/gst-plugins-rs/blob/master/meson.build.

I'm not a Rust expert at all, so please correct me!

The idea is as follow:
- Your main project uses Meson, it can build C/C++ libraries and subprojects.
- Bundle all your rust projects into a single Cargo project and let cargo handle deps between them.
- Put your rust project as a Meson subproject, and write a meson.build file for it.
- The meson.build will use a custom_target() to invoke cargo (or a wrapper script around it) and is always_stale because we trust cargo to not do useless rebuilds.
- Cargo uses pkg-config to find non-rust external dependencies. If some of those deps are built by the meson (sub)project, we can point PKG_CONFIG_PATH to Meson's dir where it generated -uninstalled.pc files for them (https://github.com/mesonbuild/meson/pull/4436).
- You have to know where Cargo built its final static libraries, ideally that part should be told by cargo by generating -uninstalled.pc (or similar) files itself.
- In your meson.build, consider those static libraries as prebuilt, and wrap them into a shared_library().

**Comment by jpakkane** [comment]
That can't really be made reliable because there are a gazillion things that can go wrong. The "correct" solution is to move to a more flatpakkey way of building and staging, [as discussed here](http://nibblestew.blogspot.com/2019/12/there-are-at-least-three-distinct.html).

**Comment by xclaesse** [comment]
> That can't really be made reliable because there are a gazillion things that can go wrong. The "correct" solution is to move to a more flatpakkey way of building and staging, [as discussed here](http://nibblestew.blogspot.com/2019/12/there-are-at-least-three-distinct.html).

The only thing you can be sure, is you won't know that until you try. And even if it turns out to be impossible, the only way you can proof cargo team that they are wrong, is by trying and showing them the issues. Until then, it's just talk.

**Comment by mensinda** [comment]
On the "mixing two build systems" front, I would like to add that we already support CMake subprojects, which work for even non-trivial CMake projects.

That being said, I don't have any experience with cargo and rust and have only skimmed through this issue and #2617. I don't know if a similar approach to CMake would work here or if the situation is completely different. However, if it is possible to get cargo to print all the required steps to download and compile dependencies at configure time (without doing any actual work), the situation should be pretty close to what we are doing with CMake right now.

I would say that the situation might be even easier since the cargo devs seem to be interested in actually supporting something like this.

But again, I have only read a few comments on these issues and have never written any rust code, so take all of this with a grain of salt.

**Comment by eli-schwartz** [comment]
@xclaesse Your "design that can work" is as far as I can tell, literally what @TingPing said 2.5 years ago:

> - Simply call Cargo making the Meson integration worthless

@mensinda meson doesn't "support two different build systems", though. :p I think it's an important distinction to make, and to keep in mind in all such discussions, that meson only supports importing definitions into native meson. By highlighting this distinction and avoiding confusing language like "support two different build systems", we can avoid ending up in discussions like "well, let's just support cargo too, by literally executing `cargo build`". And so far, that's what all previous discussions have been about, so it is *quite* important to avoid that rabbit hole again.

As for:

> since the cargo devs seem to be interested in actually supporting something like this.

Do you have a citation for this? My impression based on this issue and the referenced PR is that cargo devs are still heavily opposed to any use of the Cargo ecosystem other than relying on the actual command `cargo build`.

**Comment by sdroege** [comment]
> Do you have a citation for this? My impression based on this issue and the referenced PR is that cargo devs are still heavily opposed to any use of the Cargo ecosystem other than relying on the actual command `cargo build`.

See https://github.com/rust-lang/cargo/issues/5579, which currently exists as an experimental feature. There was nobody taking up the work and the cargo team is too understaffed to bring it to completion though, so it's currently under [discussion](https://github.com/rust-lang/cargo/issues/7614) to remove it again as it's not useful yet in its current form. The interest and understanding that it's useful is there, but someone has to do the work.

**Comment by xclaesse** [comment]
> @xclaesse Your "design that can work" is as far as I can tell, literally what @TingPing said 2.5 years ago:
> 
> > * Simply call Cargo making the Meson integration worthless

Yes indeed, with one big difference, I think we should rely on uninstalled.pc files when rust code depends on meson code.

Ideally cargo should also generate uninstalled.pc files in a configure step so we can use them in meson to find libraries inside the build dir of cargo.

The more I think about it, the more I realise that uninstalled.pc is the build system agnostic way of serializing internal dependencies structure.

**Comment by eli-schwartz** [comment]
@sdroege Interesting, so I guess there could be hope in the future.

@xclaesse "generate -uninstalled.pc" is not some magical solution to all ills, or whatever. Its use is kind of by definition only to cover the same use case as detecting system installed dependencies *outside* of a build system, except without running make/ninja install, for the strict case of a developer wanting to make compile/test cycles faster without using sysroots or installing the thing they are testing.

None of this remotely helps communication *between* build systems, since you cannot get at the commands needed to compile a project, and you cannot know when a dependency needs to be rebuilt.

**Comment by xclaesse** [comment]
@eli-schwartz nothing prevents us from using something for a use-case that it was not originally made for.

The command to compile a rust project, and rebuild it, is well known: cargo.

Howerver, remains one issue that the user needs to know which of the uninstalled.pc files cargo used, to set the dependency. Otherwise there is nothing telling ninja to build those libs before invoking cargo. Ideally that's something cargo should tell us in a configuration step.

**Comment by eli-schwartz** [comment]
And the command to compile, and recompile, an autotools project is also well known: 'make'. Let's add support for building autotools custom_target()/subproject()s next!

Advantage: autotools projects, unlike cargo projects, actually produce .pc files today.

(Problem: both cargo and make result in meson not being able to know when a project is up to date, so that gets proxied out to an external build system that gets run every time, not just when needed, and meson cannot tell the external build system when to force-remake something, nor have the external build system tell it when a meson artifact is out of date save by actually running said external build system and checking *if* the artifacts were updated.)

**Comment by xclaesse** [comment]
> And the command to compile, and recompile, an autotools project is also well known: 'make'. Let's add support for building autotools custom_target()/subproject()s next!

I totally agree we should make it for other build systems too, cargo is not special. I worked on that a couple years ago: https://github.com/mesonbuild/meson/pull/4321. In used a different idea, though.

> Advantage: autotools projects, unlike cargo projects, actually produce .pc files today.

Yes, but sadly only a few generate -uninstalled.pc files. But I guess it's easier to patch them to generate the uninstalled pc file than to write a whole meson build definition.

> (Problem: both cargo and make result in meson not being able to know when a project is up to date, so that gets proxied out to an external build system that gets run every time, not just when needed, and meson cannot tell the external build system when to force-remake something, nor have the external build system tell it when a meson artifact is out of date save by actually running said external build system and checking _if_ the artifacts were updated.)

That's not really a problem. Meson just always invoke external build systems and rely on them to be no-op if nothing changed. In practice it's good enough.

**Comment by eli-schwartz** [comment]
Is it indeed good enough? A no-op build should complete quickly and preferably without verbose output, and that's *certainly* not the case for the average autotools project (on both counts).

Waiting on cargo to provide an introspection method like the one the current cmake module relies on, would ensure meson could integrate the necessary commands natively into build.ninja and achieve both those goals.

**Comment by xclaesse** [comment]
> Is it indeed good enough? A no-op build should complete quickly and preferably without verbose output, and that's _certainly_ not the case for the average autotools project (on both counts).

Well, if that's an issue then that project certainly should consider porting to meson/cmake. We can't do miracles.

> Waiting on cargo to provide an introspection method like the one the current cmake module relies on, would ensure meson could integrate the necessary commands natively into build.ninja and achieve both those goals.

That probably means waiting a few decades, let's be honest, that just won't magically happen. And even that has serious issues, I'm not convinced at all the current cmake module way is better, totally fail at cross build, you still have the same issues to provide meson deps to cmake, etc.

**Comment by xclaesse** [comment]
TIL: Cargo actually generate depfiles and @nirbheek's meson module was already using them to avoid invoking cargo when not needed (with `build_always_stale`). So that's a problem already solved.

**Comment by amCap1712** [comment]
How can I use rust libraries with a project using meson build system and are there any examples to do so ?  I am willing to use a script and a combination of both meson and cargo if needed.

**Comment by michaelbrockus** [comment]
I have found some documentation for crates.io web API. Hopefully this would help with adding support for Cargo crates.

Relevant document [here](https://doc.rust-lang.org/cargo/reference/registries.html#web-api).

**Comment by dcbaker** [comment]
I'm working on the 'reimplement cargo' route, like the way we handle cmake. It's coming up on my list of things to get back to.

**Comment by bogen85** [comment]
This issue has been open for over 5 years... I would like to see meson support external crates with Rust. On Fedora/RHEL/RHEL-derivatives there are many system packages for rust crates. I've found no clear way to even specify they be used in a meson project. Currently I'm still having to use Cargo instead of Meson, but I'd prefer meson as that is what I use for other languages.

Cargo is great, but Rust's over dependence on having to use it exclusively as the only build tool (??) makes Rust not feel as first class to me... I hoping that Meson can help Rust feel more first class as far as project building. True, thinks like AdaMake are baked into GNAT (Ada Tooling...) but at least AdaMake lets me use Ada libraries installed in system packages... I know, this is a discussion about Rust support in Meson... Which is difficult due to tunnel vision in Rust development caused by over dependence on Cargo exclusively for anything other than simple hello world program examples...

**Comment by bogen85** [comment]
https://bugzilla.redhat.com/show_bug.cgi?id=1920959 that bug is closed due to Fedora 34 being end of life, but it highlights what I believe is at the core of the problem here... Rust's crate ecosystem does not fit well into established build systems (fedora system builds, meson, etc). And trying to make it fit is very painful. One "should" be able to install all the dependencies for a programming project just using system packages. (Yes, that might mean not being able to use some deps you would prefer to use, and having to cope with just being used what is available as an already installed packages). Rust crate ecosystem (which by default is tied to cargo) needs to updated/improved to "play well with" others.... Because right now it does not... Hopefully that is addressed so that is not such a pain to get proper support into Meson. Cargo is overly dependent on an Internet connection... And the pulling in of unvetted dependencies... That in and of it self makes Rust painful to use for many software development projects. While problems make still trckle down, at least system packages have a lot more vetting applied to them.

Even when Meson gets proper external crate support, if it has to download deps off of github that is no good in my opinion. One needs to be able to use what they have already installed on their local system.

**Comment by eli-schwartz** [comment]
There are two factors here:
- integrating well with system library dependencies (C ABI, probably written in C)
- resolving rust rlib dependencies


Ideally both would work in packaged system dependencies form. In practice rust tooling (especially cargo) doesn't really "want" to do rlibs as anything other than built in vendored form. I don't know what the distribution system for that is supposed to look like, but perhaps we can invent one.

...


I do know that @dcbaker is working on this and is very motivated to do so as having a better crates story for Meson is a blocker for Mesa. The current mechanism being proposed will likely involve supporting only crates that don't use build.rs, which can be robustly translated into meson.build rules; I think the main blocker for this is waiting for python 3.11 and then figuring out the tomllib dependency for parsing Cargo.toml.

That would allow vendoring crates as subprojects, building them via meson instead of cargo, and linking them to C libraries provided as subprojects which cargo can't know about, which is a definite and powerful improvement over the state of the art.

**Comment by dcbaker** [comment]
Yeah. I am still working toward being able to build rust crates in meson, including by vendoring the rust source in the meson produced dist tarball, as @eli-schwartz said. I've been working hard on paying down technical debt, as it's become a pain point in getting this working.

There are a bunch of issues with using rust libs as system dependencies:
1. there is no stable dynamic library ABI, which means that every change in the Rust compiler potentially requires rebuilding all of the rust packages on a system. Rust also heavily relies on templating, which makes dynamic libraries even more complicated, IIUC.
2. cargo doesn't support installing rlibs (which are static) system wide
3. pkg-config doesn't really support non-C-ABI dependencies @kaniini and I have talked about this a bit, and this is very much fixable

**Comment by sdroege** [comment]
> 1. there is no stable dynamic library ABI, which means that every change in the Rust compiler potentially requires rebuilding all of the rust packages on a system

(Also affecting 2.) Symbol names get a hash added to them that includes, among other things, the versions of all the crate's dependencies, the selected feature flags and IIRC also parts of the code related to the symbol. So changing any version of anything (including the compiler) or changing feature flags will require rebuilding everything on top of it.

You can set this hash yourself with some compiler flags IIRC but then you're really on your own with ensuring ABI compatibility.

If you want to do dynamic linking with Rust at this point you really need to build your whole set of libraries/binaries in one go.

> Rust also heavily relies on templating, which makes dynamic libraries even more complicated, IIUC.

Not very different from C++ with header-only libraries though, however the generic code is part of the rlibs in non-source form and incompatible between compiler versions.

----

Linux distros (Fedora/Debian) are shipping Rust library crates in the -dev/-devel packages as source code for these reasons, and binaries statically link them in (those distro packages are basically used as a different way of vendoring / a separate cargo crates registry).

@dcbaker I assume you're aware of all that, just adding this extra information here because your comment makes the challenges sound simpler than they actually are in that regard :)

**Comment by LingMan** [comment]
@bogen85 Cargo doesn't need a network connection if you just [tell it where to find the crates it needs locally](https://doc.rust-lang.org/cargo/reference/source-replacement.html). From the looks of it, Fedora rust packages build up a directory source at `/usr/share/cargo/registry`. From that link above: If you want to limit yourself to those, simply create a `.cargo/config.toml` with this content:
```toml
[source.my-local-fedora-source]
directory = "/usr/share/cargo/registry"

[source.crates-io]
replace-with = "my-local-fedora-source"

[net]
offline = true   
```

Untested and I don't have a Fedora system, so this might be a bit broken but should work in principle.
For a larger organization creating a vetted private crate repository would of course be a better option than a local dir.

**Comment by bogen85** [comment]
@LingMan 
> Untested and I don't have a Fedora system, so this might be a bit broken but should work in principle. For a larger organization creating a vetted private crate repository would of course be a better option than a local dir.

I tried this with a few crates in a simple project. It seems to work as intended. There is no reaching out to the Internet to do a build from scratch starting out with an empty `~/.cargo`
Thanks!

Not every crate exists obviously but there are over 10K rust packages in Fedora 36...

Multiple variants exist for many crates:
```
rust-zmq+default-devel.noarch : High-level bindings to the zeromq library
rust-zmq+zmq_has-devel.noarch : High-level bindings to the zeromq library
rust-zmq-devel.noarch : High-level bindings to the zeromq library
rust-zmq-sys+default-devel.noarch : Low-level bindings to the zeromq library
rust-zmq-sys-devel.noarch : Low-level bindings to the zeromq library
```

In many cases, all the variants for a given crate can be installed at the same time, and I'm not completely sure what is up with that yet.

**Comment by carbotaniuman** [comment]
Is there any workaround/progress for this? I want to build some multi language code with this, but this is a major blocker.

**Comment by eli-schwartz** [comment]
@carbotaniuman Yes, the recent meson release can automatically convert a Cargo.toml into a meson subproject.

https://mesonbuild.com/Release-notes-for-1-3-0.html#automatic-fallback-to-cmake-and-cargo-subproject

This does not yet work for all cases, there are pending improvements to handle cargo features correctly as well as to automatically generate wraps by contacting https://crates.io when an existing subproject has a crate dependency which we don't have locally yet. Please follow #12363 for more details (and feel free to test out this experimental PR branch).

We are anxious to land it for meson 1.4.0 as there are several interested consumers that need it (and have been the driving force behind the recent focus on getting rust crates to work).

**Comment by bonzini** [comment]
As an update, the required changes are something like this:
* #14880 [1.10.0]
* #15113 [1.10.0]  
* Allow parsing toplevel Cargo.toml, i.e. without needing a subproject
  * #14906 [1.10.0]
  * #15158 [1.11.0]
* Build targets from toplevel Cargo.toml:
  * #14921 [1.10.0]
  * #15223 [1.11.0]
* Related:
  * #15404, #14952 [cross compilation support, 1.12.0)
  * #15837 [environment variables]

The above does not yet cover dev-dependencies and tests, though that could be the next step. As of 1.11.0 it is possible to build a binary and all its dependent crates with just this:

```
cargo = import('rust').workspace()
cargo.package().executable(install: true)
```

In the meanwhile, Meson also grew support for more of the Rust toolchain; doctests, in particular, are a thorn on the side of projects that just build a staticlib with cargo and link it into C, but can be run easily with Meson.

Crates that use a `build.rs` still need to translate it to handwritten meson. From the top 150 crates there are about 20: proc-macro2, quote, libc, getrandom, serde, serde_json, parking_lot_core, num-traits, rustix, anyhow, crossbeam-utils, memoffset, rustls, ahash, ~~typenum~~, nix, ~~httparse~~, crc32fast, zerocopy.

**Comment by v1993** [comment]
Today I've tested git version of meson and was able to easily build a GTK "Hello, world!" program using rust.workspace. This is excellent news! However, it seems like it's impossible to set environment variables for your project and thus pass build time options to it (other than features)? This likely could be worked around with `configure_file` for the time being, but it is something that works perfectly fine when wrapping `cargo` with `custom_target`.

P.S.: @bonzini, it looks like the note at the bottom was left forgotten when text referencing it was removed.

**Comment by bonzini** [comment]
> However, it seems like it's impossible to set environment variables for your project and thus pass build time options to it (other than features)?

I would use features if code needs to be compiled out at the syntactic level, and otherwise  `configure_file` to generate consts.  Meson tends to shun environment variables. :)

You could also use something like `#[cfg(MESON)]` and `rust_args: ['--cfg', 'MESON']` (or `add_project_arguments`) to choose between a version using `include!` of the `configure_file`, and a version using `std::env!`.

**Comment by v1993** [comment]
I see, although it seems like using `configure_file` has additional issues: you either run into a chicken and egg problem (how do you tell rustc where to look for the configured file?) or have to place it in a predefined location, which e.g. doesn't play well with having multiple build directories. I guess the real solution would be to use `--env-set`, but it's sadly still nightly exclusive, seemingly without much movement these days to change that.

**Comment by bonzini** [comment]
We could wrap non-nightly Rust to translate --env-set into environment variables. In fact Meson does it already for rustdoc...

**Comment by v1993** [comment]
> We could wrap non-nightly Rust to translate --env-set into environment variables. In fact Meson does it already for rustdoc...

Current version of #10030 (needs rebasing) appears to do exactly that, although given its history I'm only somewhat hopeful that it will be accepted even in such form.

**Comment by bonzini** [comment]
At this point all that Meson can add is the support for environment variables, which is in progress, and possibly put some of the common crates in wrapdb. I am going to close this once #15837 is in and #15944 is fixed.
